### PR TITLE
[Draft] Qualcomm AI Engine Direct - Support kv_cached llama2 model

### PR DESCRIPTION
**Pull request by shewu-quic**
### Summary
- Support static kv_cached llama2 model
  - We reference AIMET jupyter notebooks and implement a [static LLAMA](https://github.com/CodeLinaro/executorch/blob/9681c0195a17952a95042c407d4d01e4e2b4bab1/examples/qualcomm/llama2/model/static_llama.py)
- Add [qnn_llama_runner](https://github.com/CodeLinaro/executorch/blob/9681c0195a17952a95042c407d4d01e4e2b4bab1/examples/qualcomm/executor_runner/qnn_llama_runner.cpp) to run static LLAMA
- Add [e2e example script](https://github.com/CodeLinaro/executorch/blob/9681c0195a17952a95042c407d4d01e4e2b4bab1/examples/qualcomm/llama2/llama.py) verified with story110M

### Notes
- In fp16 mode, the model can be compiled and executed on the device to obtain accurate results. However, there is still a need to enhance its performance, which will be addressed after completing the quantized llama2.
- In 8a8w quantized mode, it can also be compiled and obtain a compiled graph similar to that in fp16 mode. However, when executed on the device, the results are not as expected.
- For now, we are going to 16 bit quantization.
- The main difference between [static LLAMA](https://github.com/CodeLinaro/executorch/blob/9681c0195a17952a95042c407d4d01e4e2b4bab1/examples/qualcomm/llama2/model/static_llama.py) and [existent examples/models/llama2](https://github.com/CodeLinaro/executorch/blob/9681c0195a17952a95042c407d4d01e4e2b4bab1/examples/models/llama2/model.py) is that we regard [kv cache as the i/o of graph](https://github.com/CodeLinaro/executorch/blob/9681c0195a17952a95042c407d4d01e4e2b4bab1/examples/qualcomm/llama2/model/static_llama.py#L50).

### Compiled graph
For now, we will fallback the following ops which are about reading and updating attention mask:
- aten_index_tensor
- aten_index_put_default
image

#### Download and export stories110M model
```
# tokenizer.model & stories110M.pt:
wget "https://huggingface.co/karpathy/tinyllamas/resolve/main/stories110M.pt"
wget "https://raw.githubusercontent.com/karpathy/llama2.c/master/tokenizer.model"

# tokenizer.bin:
python -m examples.models.llama2.tokenizer.tokenizer -t tokenizer.model -o tokenizer.bin

# params.json:
echo '{"dim": 768, "multiple_of": 32, "n_heads": 12, "n_layers": 12, "norm_eps": 1e-05, "vocab_size": 32000}' > params.json
```
### Run e2e example script
```
# fp16:
python examples/qualcomm/llama2/llama.py -a xxx -b build_android -s xxx -m SM8650 -F --checkpoint stories110M --params params.json --tokenizer_bin tokenizer.bin --prompt Once
# quant:
python examples/qualcomm/llama2/llama.py -a xxx -b build_android -s xxx -m SM8650 --ptq 8a8w --tokenizer_model tokenizer.model --checkpoint stories110M --params params.json --tokenizer_bin tokenizer.bin --prompt Once
```

### CODE REVIEW

File: backends/qualcomm/partition/common_defs.py [hunk 1]
```diff
@@ -11,8 +11,9 @@
 not_supported_operator = [
     exir_ops.edge.aten.arange.start_step,
     exir_ops.edge.aten.clone.default,
-    exir_ops.edge.aten.index.Tensor,
     exir_ops.edge.aten.full.default,
+    exir_ops.edge.aten.index.Tensor,
+    exir_ops.edge.aten.index_put.default,
```
Comment (cccclai) [inline backends/qualcomm/partition/common_defs.py:16]: Maybe these two lines can be landed seperately? It's not supported anyway and we can decouple it from this large PR

File: backends/qualcomm/quantizer/utils.py [hunk 2]
```diff
@@ -119,15 +119,19 @@ def annotate_binary(node: Node, quantization_config: QuantizationConfig) -> None
...
+    output_act_qspec = (
+        quantization_config.output_activation
+        if node.meta["val"].dtype == torch.float32
+        else None
+    )
 
     input_qspec_map = {}
     input_act0 = node.args[0]
-    if isinstance(input_act0, Node):
+    if isinstance(input_act0, Node) and input_act0.meta["val"].dtype == torch.float32:
```
Comment (cccclai) [inline backends/qualcomm/quantizer/utils.py:130]: We can have a seperate PR for this line too. https://github.com/pytorch/executorch/pull/2957 has a few more checks that we may need.

File: backends/qualcomm/scripts/build.sh [hunk 3]
```diff
@@ -71,6 +71,7 @@ if [ "$BUILD_AARCH64" = true ]; then
         -DCMAKE_INSTALL_PREFIX=$BUILD_ROOT \
         -DEXECUTORCH_BUILD_QNN=ON \
         -DEXECUTORCH_BUILD_SDK=ON \
+        -DEXECUTORCH_BUILD_EXTENSION_MODULE=ON \
```
Comment (cccclai) [inline backends/qualcomm/scripts/build.sh:74]: Same here, a seperate PR

File: backends/qualcomm/tests/models.py [hunk 4]
```diff
@@ -29,7 +29,7 @@ def __init__(self):
         super().__init__()
 
     def forward(self, x):
-        return 10.0 + x
+        return 10 + x
```
Comment (cccclai) [inline backends/qualcomm/tests/models.py:32]: What's this for? Is it a different test case?

File: examples/qualcomm/scripts/utils.py [hunk 5]
```diff
@@ -212,25 +228,39 @@ def build_executorch_binary(
...
-            ),
-            extract_delegate_segments=True,
-        )
+
+    executorch_config = ExecutorchBackendConfig(
+        extract_constant_segment=False,
+        # For shared buffer, user must pass the memory address
+        # which is allocated by RPC memory to executor runner.
+        # Therefore, won't want to pre-allocate
+        # by memory manager in runtime.
```
Comment (cccclai) [inline examples/qualcomm/scripts/utils.py:237]: This seems like an optmization opportunity and we can add them later? For the alpha release, maybe let's get a functional version, and improve it step by step. What do you think?

File: examples/qualcomm/llama2/runner/runner.h [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+  bool is_loaded() const;
+  Error load();
+  Error generate(
+      const std::string& prompt,
+      int32_t seq_len,
+      ManagedTensor& managed_atten_mask,
+      ManagedTensor& managed_k_cache,
+      ManagedTensor& managed_v_cache,
+      ManagedTensor& managed_kv_mask,
```
Comment (cccclai) [inline examples/qualcomm/llama2/runner/runner.h:70]: It seems to me the main change is the model input, is it correct?

File: examples/qualcomm/llama2/model/static_llama.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+                k_cache=k,
+                v_cache=v,
+                kv_mask=kv_mask,
+            )
+            output_k_cache.append(
+                k.view(self.max_batch_size, self.max_seq_len, self.dim)
+            )
+            output_v_cache.append(
+                v.view(self.max_batch_size, self.max_seq_len, self.dim)
+            )
```
Comment (cccclai) [inline examples/qualcomm/llama2/model/static_llama.py:179]: Aren't we having a dynamic shape for `output_k_cache` here?

**Review by cccclai — COMMENTED** [review]
Looks great! I think there are many optimization we can leverage in this pr. Just wonder if we can decouple some minor changes when getting the end to end working.

Also reminder we probably need to cherry-pick

**Comment by cccclai** [comment]
Also curious what visualization tool you're using

File: backends/qualcomm/partition/common_defs.py (hunk 1 above)
Comment (haowhsu-quic) [reply backends/qualcomm/partition/common_defs.py:16]: I think we still need to have it. We need `aten.index.Tensor` for slicing kv_cache and feed to individual LlamaAttention layer. `aten.index_put` is used to update kv_mask after finishing each inference.
These two operators are not supported yet by qualcomm backend, we'll need partitioner to identify them and make them fallback to CPU.

File: backends/qualcomm/quantizer/utils.py (hunk 2 above)
Comment (haowhsu-quic) [reply backends/qualcomm/quantizer/utils.py:130]: Yes, this looks like a common issue.

File: backends/qualcomm/scripts/build.sh (hunk 3 above)
Comment (haowhsu-quic) [reply backends/qualcomm/scripts/build.sh:74]: This is related to `qnn_llama_runner`, I think we still need to land it together.

File: examples/qualcomm/llama2/runner/runner.cpp [hunk 8]
```diff
@@ -0,0 +1,386 @@
+/*
```
Comment (digantdesai) [inline examples/qualcomm/llama2/runner/runner.cpp:1]: curious if we can share this file with `models/llama2`? Just easier to maintain.

File: backends/qualcomm/tests/models.py (hunk 4 above)
Comment (haowhsu-quic) [reply backends/qualcomm/tests/models.py:32]: It's a minor fix when checking if integer type addition works. Here changes the constant to be integer type to match the function title `AddConstantLong`.

File: examples/qualcomm/scripts/utils.py (hunk 5 above)
Comment (haowhsu-quic) [reply examples/qualcomm/scripts/utils.py:237]: The shared buffer has already been done by https://github.com/pytorch/executorch/pull/2531. Shared buffer mechanism is enabled for the rest of examples but not llama now (we'll have it in the near future, need changes in `qnn_llama_runner`).
For now, we still need this refactored code for other examples being functional as usual.

File: examples/qualcomm/llama2/runner/runner.h (hunk 6 above)
Comment (haowhsu-quic) [reply examples/qualcomm/llama2/runner/runner.h:70]: Yes, to make kv_cache be static shape.

**Comment by shewu-quic** [comment]
> Also curious what visualization tool you're using

I think we use FxGraphDrawer to visualize the graph module
https://github.com/pytorch/executorch/blob/d761f99f6fc952975835f4807a12319c121c4b90/backends/qualcomm/utils/utils.py#L136

File: examples/qualcomm/llama2/model/static_llama.py (hunk 7 above)
Comment (haowhsu-quic) [reply examples/qualcomm/llama2/model/static_llama.py:179]: No, we're concatenating kv_cache calculated from all attention layers:

- Input shape of k_cache: (max_batch_size, n_layers, max_seq_len, embedding_dim)
- Slice for every attention layer: (max_batch_size, max_seq_len, embedding_dim)

We could concatenate them back without runtime shape changing.

Comment (cccclai) [reply examples/qualcomm/llama2/model/static_llama.py:179]: hmm probably need to check the graph, I just feel like `output_v_cache` shape is changing when iterating the layer, because we keep appending it

Comment (haowhsu-quic) [reply examples/qualcomm/llama2/model/static_llama.py:179]: The list append method will be changed after `torch.export`. The output kv_caches from each attention layer are connected directly to final `concat` operator in `LlamaModel.forward`.

File: backends/qualcomm/scripts/build.sh (hunk 3 above)
Comment (haowhsu-quic) [reply backends/qualcomm/scripts/build.sh:74]: Our custom runner ('llama2/runner/runner') invokes methods inside `extension/module/module.cpp`. We could get rid of it if required.

File: backends/qualcomm/partition/common_defs.py (hunk 1 above)
Comment (haowhsu-quic) [reply backends/qualcomm/partition/common_defs.py:16]: I see, we'll follow your suggestion to split PRs.

**Comment by shewu-quic** [comment]
> I'm trying to repro on my side. What QNN library version did you use? The error message on my side is
> 
> ```
> [ERROR] [Qnn ExecuTorch]: initial_sequencer_dp.cc:160:ERROR:A single op, "q::Concat" (Op ID: 315c00000086e), requires 0xa00000 bytes of TCM, which is greater than the TCM size of 0x800000!
> 
> [ERROR] [Qnn ExecuTorch]: initial_sequencer_dp.cc:167:ERROR:The name of the failing op before optimization is: "q::QNN_Reshape" (Op ID: 86e).
> 
> [ERROR] [Qnn ExecuTorch]: QnnDsp <E> "aten_view_copy_default_423" generated: Requires 0xa00000 bytes of TCM, which is greater than the TCM size of 0x800000!
> 
> [ERROR] [Qnn ExecuTorch]: QnnDsp <E> RouterX86 graph prepare failed 13
> 
> [ERROR] [Qnn ExecuTorch]: QnnDsp <E> Failed to finalize graph (id: 2) with err 1002
> 
> [ERROR] [Qnn ExecuTorch]: Failed to finalize Qnn Graph with error: 1002
> ```

I use QNN 2.20 and can reproduce on SM8475 from my side.

**Comment by cccclai** [comment]
I was using qnn 2.19 and just switch to 2.20. I'm using SM8450 on my side

**Comment by cccclai** [comment]
I was able to repro the fp version on my side, but for the 8a8w version, I hit model loading error
```
[ERROR] [Qnn ExecuTorch]:  <E> Skel failed to process context binary.
[ERROR] [Qnn ExecuTorch]:  <E> Context create from binary failed for deviceId 0 coreId 0 pdId 0 err 5005
[ERROR] [Qnn ExecuTorch]:  <E> Fail to create context from binary with err 5005
[WARNING] [Qnn ExecuTorch]:  <W> sg_stubPtr is not null, skip loadRemoteSymbols
[ERROR] [Qnn ExecuTorch]:  <E> Failed to create context from binary with err 0x138d
[ERROR] [Qnn ExecuTorch]: Can't create context from binary. Error 5005.
```
is it expected?

**Comment by shewu-quic** [comment]
> I was able to repro the fp version on my side, but for the 8a8w version, I hit model loading error
> 
> ```
> [ERROR] [Qnn ExecuTorch]:  <E> Skel failed to process context binary.
> [ERROR] [Qnn ExecuTorch]:  <E> Context create from binary failed for deviceId 0 coreId 0 pdId 0 err 5005
> [ERROR] [Qnn ExecuTorch]:  <E> Fail to create context from binary with err 5005
> [WARNING] [Qnn ExecuTorch]:  <W> sg_stubPtr is not null, skip loadRemoteSymbols
> [ERROR] [Qnn ExecuTorch]:  <E> Failed to create context from binary with err 0x138d
> [ERROR] [Qnn ExecuTorch]: Can't create context from binary. Error 5005.
> ```
> 
> is it the same issue you observe from your side?

No. For 8a8w, we could get the compiled graph which is the same as that in fp16. 
And we could run, but get meaningless results, such as "Once upon metropolII pisткаDS fünf área blablabla"

**Comment by cccclai** [comment]
turns out I forget the -ptq flag...

what does the performance look like from your side? From the log output, seems like 1-2 toks/s for fp and 0.6 toks/s. Did I miss something?

**Comment by shewu-quic** [comment]
> turns out I forget the -ptq flag...I can repro both fp and 8a8w now.
> 
> what does the performance look like from your side? From the log output, seems like 1-2 toks/s for fp and 0.6 toks/s. Did I miss something?

Great! We can start to align each other's results.
Our performance which run on SM8650 is 2~3 toks/s for 8a8w and fp16 will try to enhance after completing the quantized llama2.

**Comment by cccclai** [comment]
2~3 toks/s for 8a8w seems still really slow - do we know which part is causing the perf regression? Is delegated part runs reasonably fast and the cpu part is too slow?

**Comment by shewu-quic** [comment]
Hi @cccclai,
We fixed 16a4w accuracy issue which is resolved by [PR 3196](https://github.com/pytorch/executorch/pull/3196).

Results[0]:
Once upon a time, there was a little girl named Lily. She loved to play outside in the sunshine. One day, she saw a big, fluffy cloud in the sky. It looked like a giant cotton candy!
Lily ran inside to tell her mommy about the cloud. "Mommy, mommy, look at the big cloud in the sky! It looks like a giant cotton candy!" she said.

**Comment by salykovaa** [comment]
> Hi @cccclai, We fixed 16a4w accuracy issue which is resolved by [PR 3196](https://github.com/pytorch/executorch/pull/3196).
> 
> Results[0]: Once upon a time, there was a little girl named Lily. She loved to play outside in the sunshine. One day, she saw a big, fluffy cloud in the sky. It looked like a giant cotton candy! Lily ran inside to tell her mommy about the cloud. "Mommy, mommy, look at the big cloud in the sky! It looks like a giant cotton candy!" she said.

Dear @shewu-quic @cccclai

Does this resolve [https://github.com/pytorch/executorch/issues/2590](https://github.com/pytorch/executorch/issues/2590)? If so, I will close the issue

**Comment by salykovaa** [comment]
Dear @shewu-quic @cccclai,

does [PR 3196](https://github.com/pytorch/executorch/pull/3196) resolve the issue [https://github.com/pytorch/executorch/issues/2590](https://github.com/pytorch/executorch/issues/2590)? If so, I will close the issue. Thank you in advance!

**Comment by cccclai** [comment]
> Dear @shewu-quic @cccclai,
> 
> does [PR 3196](https://github.com/pytorch/executorch/pull/3196) resolve the issue #2590? If so, I will close the issue. Thank you in advance!

Thanks for the update and sending the fix! Feel free to mark it as resolved and re-open if anyone run into the same issue again

**Comment by chiwwang** [comment]
Rebased as https://github.com/pytorch/executorch/pull/3656/

**Comment by chiwwang** [comment]
Please see https://github.com/pytorch/executorch/pull/4142 instead.
